### PR TITLE
KDTree class.

### PR DESCRIPTION
**Pull request by luca-heltai**
Add `KDTree` class. `KDTree` uses nanoflann internally to achieve `log(N)` time search when querying for nearest neighbors, and points within balls. 

A followup PR will make use of this in `find_active_cell_around_point`, and `find_closest_vertex` to accelerate their wall-time. 

#4704

### CODE REVIEW

File: doc/news/changes/major/20170813LucaHeltai [hunk 1]
```diff
@@ -0,0 +1,7 @@
+New: A new KDTreeDistance class has been added that interfaces
+to the nanoflann library (https://github.com/jlblancoc/nanoflann).
+This class can be used to extract nearest neighbour information 
+on collection of points, query for the closes points to a target
```
Comment (davydden) [inline doc/news/changes/major/20170813LucaHeltai]: Closest

File: doc/news/changes/major/20170813LucaHeltai [hunk 2]
```diff
@@ -0,0 +1,7 @@
+New: A new KDTreeDistance class has been added that interfaces
+to the nanoflann library (https://github.com/jlblancoc/nanoflann).
```
Comment (davydden) [inline doc/news/changes/major/20170813LucaHeltai]: I don't think u need "to" here

File: include/deal.II/numerics/kdtree_distance.h [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * A wrapper for the nanoflann library, used to compute the distance from a
+ * collection of points, and to efficiently return nearest neighbors to a
+ * target point. This function uses nanoflann to efficiently partition the
+ * space in a tree. The cost of each query is then roughly of order log(n),
+ * where n is the number of points stored in this class.
+ *
+ * The wrapper provides methods that give access to some of thefunctionalities
```
Comment (davydden) [inline include/deal.II/numerics/kdtree_distance.h]: Space missing

File: include/deal.II/numerics/kdtree_distance.h [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+ * of the nanoflann library, like searching the n nearest neighbors, or
+ * searching the points that fall within a raidius of a target point.
+ */
+template<int dim>
+class KDTreeDistance
```
Comment (davydden) [inline include/deal.II/numerics/kdtree_distance.h]: Why is it called KD?

File: include/deal.II/numerics/kdtree_distance.h [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+     * A typedef used by nanoflann.
+     */
+    typedef double coord_t;
+
+
+    /**
+     * Reference to the vector of points from which we want to compute
+     * the distance.
+     */
+    const std::vector<Point<dim> > &points; //!< A const ref to the data set origin
```
Comment (gassmoeller) [inline include/deal.II/numerics/kdtree_distance.h]: Do you need the comment after the variable name?

File: include/deal.II/numerics/kdtree_distance.h [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+
+
+    /**
+     * The constrcutor needs the data set source.
```
Comment (gassmoeller) [inline include/deal.II/numerics/kdtree_distance.h]: Also `data set source` is likely some internal name? You do not use it elsewhere in this documentation, maybe `The constructor needs the vector of points from which we want to build the tree`?

File: include/deal.II/numerics/kdtree_distance.h (hunk 4 above)
Comment (luca-heltai) [inline include/deal.II/numerics/kdtree_distance.h]: https://en.wikipedia.org/wiki/K-d_tree

I'll add a comment to the documentation.

File: source/numerics/kdtree_distance.cc [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+std::vector<std::pair<unsigned int, double> > KDTreeDistance<dim>::get_closest_points(const Point<dim> &target,
+    const unsigned int n_points) const
+{
+  Assert(adaptor, ExcNotInitialized());
+  Assert(kdtree, ExcInternalError());
+  std::vector<unsigned int> indices(n_points);
+  std::vector<double> distances(n_points);
+  std::vector<std::pair<unsigned int, double> > matches(n_points);
+
+  kdtree->knnSearch(&target[0], n_points, &indices[0], &distances[0]);
```
Comment (luca-heltai) [inline source/numerics/kdtree_distance.cc]: yes it does.

**Pull request merged**